DSM Demo and Documentation (#1981)
BabylonJS/Documentation: content/How_To/events/How_to_use_DeviceSourceManager.md

---
PG_TITLE: How to use the DeviceSourceManager
---

# DeviceSourceManager

> The [DeviceSourceManager](https://doc.babylonjs.com/api/classes/babylon.devicesourcemanager) is a class that will manage the connections for various user input devices and provide methods of querying those devices for their current state.

Reviewer: Maybe also expand a bit on the scope of the supported input devices? what types of devices can a user reasonably expect to work seamlessly using this device source manager?

Author: Good call, I'll get that added.

Author: Expanded section with supported devices

To use this, first create a DeviceSourceManager.  You will need to provide an engine object.
```javascript

> var deviceSourceManager = new BABYLON.deviceSourceManager(scene.getEngine());

Reviewer 2: ```suggestion
var deviceSourceManager = new BABYLON.DeviceSourceManager(scene.getEngine());
```

Author: fixed

```

Within your scene's render/game loop, you can query the DeviceSourceManager for the current state of a specific input.  First, you will need to get the DeviceSource object.  With this object, you can then query for a specific input's status.

In Javascript, you should use the following approach
```javascript

> // If the device has been registered in the DSM

Reviewer 2: I would recommend spelling everything out in documentation unless you explicitly say what the initialism means.
```suggestion
// If the device has been registered in the device source manager
```

Author: Changed

if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox)) {
    // And the A button was pressed

>     if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox).getInput(BABYLON.XboxInput.A) ==  1) {

Reviewer 2: nit:
```suggestion
    if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox).getInput(BABYLON.XboxInput.A) == 1) {
```

Author: fixed

This class supports several methods of input:
- **Keyboard** *(DeviceType: BABYLON.DeviceType.Keyboard, Inputs: number)*
- **Mouse** *(DeviceType: BABYLON.DeviceType.Mouse, Inputs: BABYLON.PointerInput)*
- **Touch** *(DeviceType: BABYLON.DeviceType.Pointer, Inputs: BABYLON.PointerInput)*
- **DualShock Gamepad** *(DeviceType: BABYLON.DeviceType.DualShock, Inputs: BABYLON.DualShockInput)*
- **Xbox Gamepad, One or 360** *(DeviceType: BABYLON.DeviceType.Xbox, Inputs: BABYLON.XboxInput)*
- **Switch Gamepad, L+R JoyCon Grip or Pro Controller** *(DeviceType: BABYLON.DeviceType.Switch, Inputs: BABYLON.SwitchInput)*
- **Generic/Other Gamepad** *(DeviceType: BABYLON.DeviceType.Generic, Inputs: number)*

To use this, first create a DeviceSourceManager.  You will need to provide an engine object.
```javascript
var deviceSourceManager = new BABYLON.deviceSourceManager(scene.getEngine());
```

Within your scene's render/game loop, you can query the DeviceSourceManager for the current state of a specific input.  First, you will need to get the DeviceSource object.  With this object, you can then query for a specific input's status.

> In Javascript, you should use the following approach

Reviewer 2: [Optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#Browser_compatibility) is mostly already available in major browsers for JavaScript, so I would just use it without saying anything about TypeScript. Maybe you can mention "optional chaining" as a way to shorten the line.

Author: Ah, okay.  I wasn't sure if got added yet.  Is there any reason why optional chaining doesn't work in the PG?

Author: Changed section

Reviewer 2: It should work? Does it not work?

Author: Nope, I've tried it and it doesn't work

Reviewer 2: Oh, it looks like the feature is behind a flag.

> From version 79: this feature is behind the Experimental JavaScript preference (needs to be set to true).

```javascript
// If the device has been registered in the DSM
if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox)) {
    // And the A button was pressed
    if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox).getInput(BABYLON.XboxInput.A) ==  1) {
        // Do something
    }
}
```

In Typescript, you can use a null-conditional operator to make it a one line call
```javascript
if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox)?.getInput(BABYLON.XboxInput.A) ==  1) {
    // Do something
}

> ```

Reviewer 2: If you are using TypeScript, it's probable you are also using modules, so you probably don't have the `BABYLON` prefix.

Author: I removed the explicit ref to Typescript

# Events and Observables
You can use the following Observables to work with identifiers for a given device
```javascript

> /*  Before a device is registered */

Reviewer 2: nit: might consider using `//` for consistency

Author: Changed

});

onBeforeDeviceDisconnectedObservable.add((device) => {
    // You can get the device type by using device.deviceType
    // You can also get the device slot (only applicable to gamepads and touch) by using device.deviceSlot
});

/*  After a device is registered */
onAfterDeviceConnectedObservable.add((device) => {
    // You can get the device type by using device.deviceType
    // You can also get the device slot (only applicable to gamepads and touch) by using device.deviceSlot
});

onAfterDeviceDisconnectedObservable.add((device) => {
    // You can get the device type by using device.deviceType
    // You can also get the device slot (only applicable to gamepads and touch) by using device.deviceSlot
});
```

For Keyboards and Pointers, you can use an event based system to get the current input and previous input when an input is activated
```javascript
deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Keyboard).onInputChangedObservable.add((device) => {
    // device.inputIndex is the activated input
    // device.currentState is the current value
    // device.previousState is the previous value (before activation)
});
```

> * [Playground Example](https://playground.babylonjs.com/#C7PM2B)

Reviewer 2: Maybe it's better to put this up top?

Author: Moved
